secrets/secrets-kubernetes/main.go

```diff
@@ -0,0 +1,178 @@
+		return "", "", secrets.ErrInvalidRequest
+	}
+
+	return parts[0], parts[1], nil
+}
+
+func parseApplicationPolicy(r *secrets.Request) (*kubeApplicationPolicy, error) {
+	rawPolicy, err := r.Context.Application.PolicyProperties()
+	if err != nil {
+		return nil, err
+	}
+	policy := &kubeApplicationPolicy{Namespace: "default"}
+	err = json.Unmarshal(rawPolicy, policy)
+	return policy, err
+}
+
+func (s *kubeSecretsServer) Get(ctx context.Context, r *secrets.Request) (*secrets.Response, error) {
+	policy, err := parseApplicationPolicy(r)
+	if err != nil {
+		return nil, fmt.Errorf("%w: invalid policy format", secrets.ErrInvalidRequest)
+	}
+	slog.Info("Get", slog.Any("application", r.Context.Application.Name), slog.Any("key", r.Name), slog.Any("impersonate", policy.Impersonate))
+
+	kubeSecretName, kubeEntryName, err := s.Locator(r.Name)
+	if err != nil {
+		return nil, fmt.Errorf("%w: invalid secret name format", secrets.ErrInvalidRequest)
+	}
+
+	kubeClient, err := kubeClientWithImpersonation(policy.Impersonate)
+	if err != nil {
+		return nil, fmt.Errorf("%w: couldn't impersonate (%s)", secrets.ErrInvalidRequest, err.Error())
+	}
+
+	kubeSecret, err := kubeClient.Secrets(policy.Namespace).Get(ctx, kubeSecretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", secrets.ErrPolicy, err.Error())
+	}
+
+	kubeEntryValue, ok := kubeSecret.Data[kubeEntryName]
+	if !ok {
+		return nil, fmt.Errorf("%w: secret not found", secrets.ErrInvalidRequest)
+	}
+
+	return &secrets.Response{
+		Secret: &secrets.SecretValue{
+			Name:         r.Name,
+			StringSecret: string(kubeEntryValue),
+			Version:      kubeSecret.ResourceVersion,
+		},
+	}, nil
+}
+
+func kubeClientWithImpersonation(role string) (clientcorev1.CoreV1Interface, error) {
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	config, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, nil).ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	if role != "" {
+		config.Impersonate.UserName = role
+	}
+
+	kubeClientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return kubeClientset.CoreV1(), nil
+}
+
+func main() {
+	var (
+		natsURL            = flag.String("nats-url", nats.DefaultURL, "Nats URL")
+		secretsBackendSeed = flag.String("backend-seed", "", "NKeys Curve Seed. Leave blank for ephemeral key")
+	)
+	flag.Parse()
+
+	slog.Info("Starting", slog.Any("nats-url", *natsURL))
+
+	s := &kubeSecretsServer{}
+
+	nc, err := nats.Connect(*natsURL)
+	if err != nil {
+		slog.Error("Couldn't setup nats client", slog.Any("error", err))
+		os.Exit(1)
+	}
+
+	errorCallback := func(_ *nats.Msg, err error) {
+		slog.Error("server error", slog.Any("error", err))
+	}
+
+	var secretsBackendKey nkeys.KeyPair
+
+	if *secretsBackendSeed != "" {
+		secretsBackendKey, err = nkeys.FromSeed([]byte(*secretsBackendSeed))
+	} else {
+		slog.Info("Creating ephemeral curve keys. DO NOT USE THIS IN PRODUCTION.")
```

Comment:
I actually like the idea of using a completely ephemeral curve key, why wouldn't we want to do that given that the generated key shouldn't need to be used elsewhere?

Is the concern that, if you use an ephemeral key, that the host that's querying the backend may be unknowingly talking to a secrets backend that's malicious?

Reply:
One downside to a just-in-time generated key as opposed to pre-generated key would be that you couldn't run multiple instances of the secret backend, because they would have different keys.

Reply:
> Is the concern that, if you use an ephemeral key, that the host that's querying the backend may be unknowingly talking to a secrets backend that's malicious?

precisely. as of now wasmcloudhost caches the backend server_xkey indefinitely, so we need to restart wasmcloudhost on secret backend key rotation.
the ⚖️  is that if we re-request server_xkey upon failure on the host, we might be talking to a rogue server. so maybe we need to look into verifying if a payload was signed with a known key ( signing keys style )

Reply:
It should be fine as long as you don't want to run more than one instance of the secrets backend. If you did want to run more than one pod for the backend then you'd have an issue if each one has an ephemeral key, since each instance would have a _different_ key, which means that hosts could have a chance of not being able to decrypt the response.

We talked about ways to mitigate this by having some sort of stateful API endpoint a host could to get a key and that backends would query to get their private key, but I think that may need to come in a later iteration of the secrets api.